Add numberFormatI18n Port to `@wordpress/i18n`

USERSATOSHI opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes #22628



<!-- In a few words, what is the PR actually doing? -->
This PR adds `numberFormatI18n` function to the `@wordpress/i18n` package that formats numbers according to the current locale settings, providing a JavaScript equivalent to WordPress's PHP `number_format_i18n` function

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
WordPress translation functions have had `number_format_i18n` for formatting numbers in a locale-appropriate way for some time. This functionality was missing from the JavaScript `@wordpress/i18n` package, making it difficult for developers to format numbers consistently with the site's locale in JavaScript/React components.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The implementation adds a `numberFormatI18n` function that:

1. **Leverages locale information from the i18n system**: Uses the `lang` property from existing locale data rather than adding dependencies on other packages
2. **Converts WordPress locales to JavaScript-compatible format**: Transforms WordPress locale format (e.g., `de_DE`) to JavaScript BCP 47 format (e.g., `de-DE`)
3. **Uses native Intl.NumberFormat**: Utilizes the browser's built-in internationalization API for consistent formatting
4. **Handles edge cases**: 
   - Falls back to `en-US` for invalid or unrecognized locales
   - Clamps decimal places to 0-20 range for Node.js compatibility
   - Supports domain-specific locale settings
  
#### Syntax
```ts
// CreateI18n
numberFormatI18n(number: number, decimals?: number, domain?: TextDomain): string

// DefaultI18n
numberFormatI18n(number: number, decimals?: number): string
```

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
Unit Tests for this function in `create-i18n` test file should suffice.

## Screenshots or screencast <!-- if applicable -->

<img width="643" height="429" alt="image" src="https://github.com/user-attachments/assets/e4ae8395-b024-41b6-a0b0-4e8a8d928215" />

USERSATOSHI committed:
feat: add numberFormatI18n port to `@wordpress/i18n`

swissspidy commented on packages/i18n/src/create-i18n.ts:
```diff
@@ -389,6 +397,44 @@ export const createI18n = < TextDomain extends string >(
+			plural_forms: 'n === 1 ? 0 : 1',
+		};
+
+		const wpLocale = localeData.lang || 'en_US';
+		let jsLocale;
+		if ( ! I18N_WP_LOCALE_REGEXP.test( wpLocale ) ) {
+			// If the locale is not in the expected format, default to 'en-US'.
+			jsLocale = 'en-US';
+		} else {
+			jsLocale = wpLocale.replace( '_', '-' );
+		}
```
Let's make sure this works for all the locales listed at https://make.wordpress.org/polyglots/teams/

I just tested them with `Intl.getCanonicalLocales()` and here are my findings:

* `de-DE-formal` are actually supported (I thought it wasn't), so we should make it work too.
* `art-xemoji` is supported too
* `pt-PT-ao90` doesn't work, and `Intl.getCanonicalLocales()` throws

So a regex is too overkill, maybe simply replacing the underscores is enough. And then perhaps `try { Intl.getCanonicalLocales() } catch ...` to prevent errors.

We should have tests for such "special" locales too.

USERSATOSHI replied:
Oh, thanks for informing, updating to reflect this.

USERSATOSHI replied:
> pt-PT-ao90 doesn't work, and Intl.getCanonicalLocales() throws 

Seems likes this is not yet supported in both browsers and Node.js Intl so either we have to make it fallback to `en-US` ( current behaviour ) or write an if statement to make it fallback to `pt-PT`.

Other than that, I have changed to use  getCanonicalLocales() to test locale and I have updated the tests to include both special locales and all the wp Locales mentioned in [make.wordpress.org/polyglots/teams](https://make.wordpress.org/polyglots/teams/)

cc: @swissspidy

USERSATOSHI committed:
refactor: switch to a more leaner locale check and added test
- added tests for special locales and all polygot locales
- switched form regex to a getCanonicalLocale() test to check locale validity

USERSATOSHI committed:
tests: fix tests according v20.19.4

swissspidy commented on packages/i18n/src/test/create-i18n.ts:
```diff
@@ -415,6 +415,439 @@ describe( 'createI18n', () => {
+		} );
+
+		// Comprehensive test for all WordPress Polyglots locales
+		test( 'comprehensive polyglots locale coverage', () => {
```
Why not use `test.each` instead of making this complicated with collecting failures etc?

USERSATOSHI replied:
Oh, I didn't know about test.each and it sadly didn't show up in my intellisense. 

updating it to use test.each

USERSATOSHI committed:
tests: refactor polygot test to use test.each

swissspidy commented on packages/i18n/src/test/create-i18n.ts:
```diff
@@ -415,6 +415,419 @@ describe( 'createI18n', () => {
+					expect( result ).toBe( expectedFormat );
+				}
+			);
+
+			test( 'should have 208 locales tested', () => {
+				expect( polyglotLocales.length ).toBe( 208 );
+			} );
```
```suggestion
```

Not a useful test, `polyglotLocales` is a static array.

USERSATOSHI committed:
tests: remove unnecessary tests

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

swissspidy commented on packages/i18n/src/create-i18n.ts:
```diff
@@ -389,6 +406,44 @@ export const createI18n = < TextDomain extends string >(
+			jsLocale = 'en-US';
+		}
+
+		/**
+		 * Range limit of `maximumFractionDigits` in Node.js is 0-20
+		 * Max limit of `maximumFractionDigits` in Browsers is 0-100 ( {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#maximumfractiondigits} )
+		 *
+		 * Hence we limit the decimals to 20 to ensure compatibility across environments.
+		 */
+		const validDecimals = Math.max( 0, Math.min( decimals, 20 ) );
```
What's the motivation for this? Is it really our responsibility to clamp this value? Seems more like the consumer's job.

USERSATOSHI replied:
The reason behind this was that during testing I found out that while Browser supported upto 100 decimals , Nodejs was limited to around 20 decimal points and anything above that it would throw an error saying invalid range.
Hence I thought of adding clamp to this to prevent this error.
